hazelcast/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java

```diff
@@ -144,36 +168,87 @@ private boolean processItem(@Nonnull T item) {
+                (r, e) -> resultQueue.add(tuple3(item, lastWatermarksAtReceiveTime, r != null ? r : e))));
         inFlightItems.merge(item, 1, Integer::sum);
         return true;
     }
 
     @Override
     public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
-        keyedWatermarkCheck(watermark);
         if (!emitFromTraverser(currentTraverser)) {
             return false;
         }
-        assert lastEmittedWm <= lastReceivedWm : "lastEmittedWm=" + lastEmittedWm + ", lastReceivedWm=" + lastReceivedWm;
+        int wmIndex = getWmIndex(watermark.key());
+
+        assert lastEmittedWms[wmIndex] <= lastReceivedWms[wmIndex]
+                : "lastEmittedWm=" + lastEmittedWms[wmIndex] + ", lastReceivedWm=" + lastReceivedWms[wmIndex];
         // Ignore a watermark that is going back. This is possible after restoring from a snapshot
         // taken in at-least-once mode.
-        if (watermark.timestamp() <= lastReceivedWm) {
+        if (watermark.timestamp() <= lastReceivedWms[wmIndex]) {
             return true;
         }
-        if (watermarkCounts.isEmpty()) {
+        if (allEmpty(watermarkCounts)) {
             if (!tryEmit(watermark)) {
                 return false;
             }
-            lastEmittedWm = watermark.timestamp();
+            lastEmittedWms[wmIndex] = watermark.timestamp();
         }
-        lastReceivedWm = watermark.timestamp();
+        // We must not mutate lastReceivedWms, because we share the instance in the inFlightItems - we would
+        // mutate the instance they have. Instead, we copy and mutate it.
+        lastReceivedWms = Arrays.copyOf(lastReceivedWms, Math.max(wmIndex + 1, lastReceivedWms.length));
+        lastReceivedWms[wmIndex] = watermark.timestamp();
         return true;
     }
 
+    /**
+     * Get the index for the given WM key. Extends the arrays, if needed.
+     */
+    @SuppressWarnings("checkstyle:MagicNumber")
+    private int getWmIndex(byte wmKey0) {
+        int wmKey = wmKey0 & 0xff; // convert the key to range 0..255
+        if (wmKeysInv.length <= wmKey) {
+            int oldLength = wmKeysInv.length;
+            int newLength = wmKey + 1;
+            wmKeysInv = Arrays.copyOf(wmKeysInv, newLength);
+            for (int i = oldLength; i < newLength; i++) {
+                wmKeysInv[i] = -1; // -1 marks unused wm key
+            }
+        }
+        int wmIndex = wmKeysInv[wmKey];
+
+        // if the wm key is seen for the 1st time, extend the arrays
+        if (wmIndex < 0) {
+            wmIndex = wmKeys.length;
+            wmKeysInv[wmIndex] = wmKey0;
+            int newLength = wmKeys.length + 1;
+
+            wmKeys = Arrays.copyOf(wmKeys, newLength);
+            wmKeys[wmIndex] = (byte) wmKey; // thanks to the cast, the key will be again in range -128 .. 127
+
+            lastReceivedWms = Arrays.copyOf(lastReceivedWms, newLength);
```

Comment:
I understand why we must copy this array, but is it necessary to copy all others? AFAIU, only this is sent via future.

Reply:
You can't change length of an array without copying.